t/apicast-policy-routing.t

```diff
@@ -0,0 +1,2083 @@
+exp => time + 3600 }, key => \$::rsa, alg => 'RS256', extra_headers=>{ kid => 'somekid' });
+"Authorization: Bearer $jwt"
+--- error_code: 200
+--- response_body
+GET / HTTP/1.1
+--- no_error_log
+[error]
+
+=== TEST 21: the rule matches a jwt claim using "!=" and the condition is true
+--- backend
+   location /transactions/oauth_authrep.xml {
+     content_by_lua_block {
+       ngx.exit(200)
+     }
+   }
+--- configuration
+{
+  "oidc": [
+    {
+      "issuer": "https://example.com/auth/realms/apicast",
+      "config": {
+        "id_token_signing_alg_values_supported": [
+          "RS256"
+        ]
+      },
+      "keys": {
+        "somekid": {
+          "pem": "-----BEGIN PUBLIC KEY-----\nMFwwDQYJKoZIhvcNAQEBBQADSwAwSAJBALClz96cDQ965ENYMfZzG+Acu25lpx2K\nNpAALBQ+catCA59us7+uLY5rjQR6SOgZpCz5PJiKNAdRPDJMXSmXqM0CAwEAAQ==\n-----END PUBLIC KEY-----"
+        }
+      }
+    }
+  ],
+  "services": [
+    {
+      "id": 42,
+      "backend_version": "oauth",
+      "backend_authentication_type": "service_token",
+      "backend_authentication_value": "token-value",
+      "proxy": {
+        "authentication_method": "oidc",
+        "oidc_issuer_endpoint": "https://example.com/auth/realms/apicast",
+        "api_backend": "http://example.com:80/",
+        "proxy_rules": [
+          {
+            "pattern": "/",
+            "http_method": "GET",
+            "metric_system_name": "hits",
+            "delta": 2
+          }
+        ],
+        "policy_chain": [
+          {
+            "name": "apicast.policy.routing",
+            "configuration": {
+              "rules": [
+                {
+                  "url": "http://test:$TEST_NGINX_SERVER_PORT",
+                  "condition": {
+                    "operations": [
+                      {
+                        "thing_to_match": "jwt_claim",
+                        "jwt_claim_name": "test_jwt_claim",
+                        "op": "!=",
+                        "value": "some_value"
+                      }
+                    ]
+                  }
+                }
+              ]
+            }
+          },
+          {
+            "name": "apicast.policy.echo"
+          },
+          {
+            "name": "apicast.policy.apicast"
+          }
+        ]
+      }
+    }
+  ]
+}
+--- upstream
+  location / {
+     content_by_lua_block {
+       ngx.say('yay, api backend');
+     }
+  }
+--- request
+GET /
+--- more_headers eval
+use Crypt::JWT qw(encode_jwt);
+my $jwt = encode_jwt(payload => {
+test_jwt_claim => 'i_dont_match',
+aud => 'the_token_audience',
+sub => 'someone',
+iss => 'https://example.com/auth/realms/apicast',
+exp => time + 3600 }, key => \$::rsa, alg => 'RS256', extra_headers=>{ kid => 'somekid' });
+"Authorization: Bearer $jwt"
+--- error_code: 200
+--- response_body
+yay, api backend
+--- no_error_log
+[error]
+
+=== TEST 22: the rule matches a jwt claim using "!=" and the condition is false
```

Comment:
I think this integration test has way too many test cases to be readable and understandable for me.
Testing all possible options and their negations can be done in lower level tests line unit tests.

I'd rather leave this focused on high level operations, how it treats the http request, what Host header it passes, how it mangles request path etc.

Reply:
Hmm I think that for each operation (`matches` , `==` ... ) it makes sense to test both the case where it evaluates to true and false. I see what you mean, this test is too verbose, but I think part of it is because sending a request with a JWT requires quite a few lines in the config and the request.

Reply:
I think it makes sense to test both cases for each operation, but probably more in a unit test rather than integration one. Integration tests should be focused on the HTTP request itself.
I get that that unit tests might be a bit different because there is no real http request, but that is fine in my opinion. If we find out we have issues with it we try to solve that.

Maybe extracting those common parts into some shared helpers or something.

Reply:
Fair enough. I think that for this particular case it's enough to test operations when they match. If we try to match a header and it works, it means that the code was able to extract the header specified correctly, which is one of the things we want to do in these integration tests. Adding an extra test with the same information but evaluating to false does not add much value.

All the operations are covered in the unit tests of the `Condition` module.

I deleted the redundant tests following that guideline. They went from 32 to 20: https://github.com/3scale/apicast/pull/976/commits/5fa61d21adc02ca235ac74ac68ad93330c6cd443